### PR TITLE
Fix type test symbol handling

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Improve type test symbols handling. Well known symbols are preserved, as is `symbol`, but unique symbols are replaced (with never), both when used as keys and as value types preventing valse positives (detected API breaks that are op-ops).

This was manually tested with the RC4 client packages, a determined to fix the issue with the new fluid handle symbol.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts [hunk 1]
```diff
@@ -26,26 +26,49 @@ type requireAssignableTo<_A extends B, B> = true;
...
-	? T
-	: {
-			[P in keyof T]: TypeOnly<T[P]>;
-	  };
+		? string
+		: T extends boolean | bigint | symbol
+			? FilterKey<T>
+			: {
+					[P in keyof T as FilterKey<P>]: TypeOnly<T[P]>;
+				};
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/typeValidator/compatibility.ts:49]: I like the new indentation better, just checking that we confirmed prettier/biome won't complain about the change? I don't remember if the auto-generated files end up getting formatted or not.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts [hunk 2]
```diff
@@ -26,26 +26,49 @@ type requireAssignableTo<_A extends B, B> = true;
  * This will strip some type branding information which ideally would be kept for stricter checking, but without it, const enums show up as breaking when unchanged.
  */
 export const typeOnly = `
-// See 'build-tools/src/type-test-generator/compatibility.ts' for more information.
+// See 'build-tools/packages/build-tools/src/typeValidator/compatibility.ts' for more information.
+type ValueOf<T> = T[keyof T];
+type WellKnownSymbols = OnlySymbols<ValueOf<typeof Symbol>>;
+type OnlySymbols<T> = T extends symbol ? T : never;
+// Omit (replace with never) a key if it is a symbol thats not a well known symbol from global Symbol.
+type FilterKey<P> = symbol extends P
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/typeValidator/compatibility.ts:34]: `symbol extends P`? What do symbols extend from? Am I understanding right that this means something like "if the key's type is so 'fundamental' that even symbols extend from it, then the key should be considered for type testing"? (The rest of the ternary(ies) I do understand).

**Review by alexvy86 — COMMENTED** [review]
Overall idea makes sense to me, but we definitely want someone more versed in Typescript ~black magic~ type handling to review. Left a few comments. 

Also since we're here, it seems line 20 needs a `)` at the end instead of a `]`.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts (hunk 2 above)
Comment (alexvy86) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts:34]: So the only case where `symbol extends P` is true is when `P` is `symbol` itself (and not any custom unique symbol), correct? In that case we want to allow the key, so use `P`. Else, use `P` if it's not a symbol at all or it is a well-known symbol, and use `never` if it's a custom unique symbol to disallow them.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts (hunk 1 above)
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts]: Our prettier ignore has `**/src/**/test/types/*.generated.ts` to cover this.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts (hunk 2 above)
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts]: I have updated the code and docs a bit to make this more clear.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts [hunk 3]
```diff
@@ -26,26 +26,49 @@ type requireAssignableTo<_A extends B, B> = true;
...
-	: T extends boolean | bigint | symbol
-	? T
-	: {
-			[P in keyof T]: TypeOnly<T[P]>;
-	  };
+		? string
+		: T extends boolean | bigint | symbol
+			? FilterKey<T>
+			: {
+					[P in keyof T as FilterKey<P>]: TypeOnly<T[P]>;
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts]: > I'm having a hard time wrapping my head around `as FilterKey<P>` here... Does it make it so keys of T which are not-well-known symbols get replaced with `never` here, and that somehow results in the key "disappearing" from the view of T produced by `TypeOnly<T>`?

Yes: that is how you remove members in mapped types.

**Review by jason-ha — APPROVED** [review]
Looks good to me.
Other improvements I am thinking about:
1. Have an import for TypeOnly instead of injecting into every file.
```typescript
import type { TypeOnly } from "@fluidframework/build-tools";
```
which would be good to keep added complexity like this from spreading.

2. Detect simple self-recursion and filter those out or replace with "type-only-recurses-here". We can rely on all of the other checks up to that point to validate. No need to check the type again and thus support recursive types.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts [hunk 4]
```diff
@@ -17,35 +17,74 @@ type requireAssignableTo<_A extends B, B> = true;
...
+		: Key; // Key is not a symbol (for example its a string or number), so leave it as is.
+/**
+ * Remove details of T which are incompatible with type testing while keeping as much as is practical.
+ *
+ * See 'build-tools/packages/build-tools/src/typeValidator/compatibility.ts' for more information.
+ */
 type TypeOnly<T> = T extends number
 	? number
 	: T extends string
-	? string
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts:33]: Maybe its not needed. Its hard to tell with the tests in a bad state for multiple other reasons, but I think we can keep string, maybe. I'll try and.

File: build-tools/packages/build-tools/src/typeValidator/compatibility.ts [hunk 5]
```diff
@@ -17,35 +17,74 @@ type requireAssignableTo<_A extends B, B> = true;
...
-	: T extends boolean | bigint | symbol
-	? T
-	: {
-			[P in keyof T]: TypeOnly<T[P]>;
-	  };
+		? string
+		: T extends boolean | bigint | symbol
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/compatibility.ts]: Done

**Pull request merged**